src/sources/generic.rs

```diff
@@ -130,16 +126,38 @@ impl<F: AsRawFd> EventSource for Generic<F> {
         Ok(())
     }
 
-    fn unregister(&mut self, poll: &mut Poll) -> std::io::Result<()> {
+    fn unregister(&mut self, poll: &mut Poll) -> crate::Result<()> {
         poll.unregister(self.file.as_raw_fd())?;
         self.token = Box::new(Token::invalid());
         Ok(())
     }
 }
 
+/// An error arising from processing events in a generic event source. Usually
+/// this will be an IO error.
+#[derive(thiserror::Error, Debug)]
+pub enum GenericError {
+    /// This variant wraps a [`std::io::Error`], which might arise from
+    /// operations on the underlying file descriptor.
+    #[error("underlying IO error")]
+    Io(#[from] std::io::Error),
+
+    /// This variant wraps any other kind of error.
+    #[error("error dispatching event")]
+    Other(#[from] Box<dyn std::error::Error + Sync + Send>),
+}
+
+impl From<nix::errno::Errno> for GenericError {
+    /// Converts a [`nix::Error`] into a wrapped version of the equivalent
+    /// [`std::io::Error`].
+    fn from(err: nix::errno::Errno) -> Self {
+        Into::<std::io::Error>::into(err).into()
+    }
+}
+
```

Comment:
Regarding the error type for `Generic`, given this source is really meant as a building block for other event sources, I wonder if it would be more convenient to instead make it possible to specify the error type, and so have a `Generic<F, E>`, with a default to `E = std::io::Error` for convenience.

That would make something like this:

```rust
struct Generic<F: AsRawFd, E=std::io::Error> {
    _error_type: std::marker::PhantomData<E>,
   /* the rest of the struct */
}

impl<F, E> EventSource for Generic<F, E> where F: AsRawFd, E: Error + Send + Sync + 'static {
    type Error = E;
    type Ret = Result<Postaction, E>;

    /* the rest of the impl */
}
```

This way, users that are fine with `std::io::Error` can just use `Generic<F>` with the default value for `E`, and users that want to customize it for their internal use still can. How does that sound?

Reply:
You know, I was just hacking around this with both the `inotify` and `gpio` crates, I think this would work quite well.

Reply:
Ah wait, this might require "generic associated types" which are still being implemented. I'll look into it.

Reply:
Nope, I was wrong, it works fine. Let me check with some non-std crates.

Reply:
Annoyingly the default type parameter doesn't actually help a lot of the time, but it's not onerous

So

```rust
Generic::new(420, Interest::READ, Mode::Level)
```

might need to become

```rust
Generic::<_>::new(420, Interest::READ, Mode::Level)
```

see https://stackoverflow.com/questions/57824584/why-rust-dont-use-default-generic-parameter-type

Reply:
We can use the same trick the stdlib uses to handle the `Alloc` parameter with collections, something like this:

```rust
struct Generic<F: AsRawFd, E=std::io::Error> {
    _error_type: std::marker::PhantomData<E>,
    fd: F
   /* the rest of the struct */
}

impl<F: AsRawFd> Generic<F, std::io::Error> {
    fn new(fd: F) -> Generic<F, std::io::Error> {
        Generic {
            fd,
            _error_type: std::marker::PhantomData,
        }
    }
    
    fn new_with_error<E>(fd: F) -> Generic<F, E> {
        Generic {
            fd,
            _error_type: std::marker::PhantomData,
        }
    }
}
```

Reply:
Nice, that worked. One more thing, should the trait bound be `Into<Box<dyn Error + etc>>` to match that on `EventSource`?

Reply:
Sounds good yes. :+1:

Reply:
Extremely happy with this change, it leads to some good gains in error handling for generic-based event sources eg. I can use anyhow to propagate informative errors up to a point where I can log them, or use my own structured errors to manage things. I deleted about 200 lines of extension trait and error mapping code in my own project.